src/train_cmd.cpp

```diff
@@ -959,12 +960,25 @@ static CommandCost CheckNewTrain(Train *original_dst, Train *dst, Train *origina
 static CommandCost CheckTrainAttachment(Train *t)
 {
 	/* No multi-part train, no need to check. */
-	if (t == nullptr || t->Next() == nullptr || !t->IsEngine()) return CommandCost();
+	if (t == nullptr || t->Next() == nullptr) return CommandCost();
 
 	/* The maximum length for a train. For each part we decrease this by one
 	 * and if the result is negative the train is simply too long. */
 	int allowed_len = _settings_game.vehicle.max_train_length * TILE_SIZE - t->gcache.cached_veh_length;
 
+	/* For free-wagon chains, check if they are within the max_train_length limit. */
+	if (!t->IsEngine()) {
+		t = t->Next();
```

Comment:
I do not know if I should call `head->InvalidateNewGRFCache();` before doing this, honestly. I hope someone can help me out there.

Reply:
All good.
* InvalidateNewGRFCache() is potato/potato here. It's only needed for the callbacks below.
* SetNext() is only changed temporary during the callback below, it's immediately restored.